# Create a simple predictive model of the target variable - "simple" meaning choose just ONE explanatory variable. 

In [182]:
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf               
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
import plotly.express as px
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score  
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
import statsmodels.api as sm
from statsmodels.tools.eval_measures import rmse
from bioinfokit.analys import stat
from bioinfokit import visuz

# DATASET 1 - AIRFARE

In [183]:
ADF = pd.read_csv(r'C:\Users\anagbhid\Documents\CourseWork_Q1\CIS508_Coursework\HandsOn3CIS_508\Airfares.csv')

In [184]:
ADF = ADF.drop(columns=[ 'S_CODE', 'S_CITY', 'E_CODE','E_CITY'])

In [185]:
ADF = pd.get_dummies(ADF, columns = ['SLOT', 'GATE', 'VACATION','SW'])

In [186]:
# from sklearn.preprocessing import MinMaxScaler
# scaler = MinMaxScaler()
# ADF_norm = pd.DataFrame(scaler.fit_transform(ADF), columns=ADF.columns)

# Question 1

# How did you choose the explanatory variable?

In [187]:
#Choosing the variable with highest correlation - "DISTANCE"
ADFcorr_matrix = ADF.corr()
ADFcorr_matrix["FARE"].sort_values(ascending=False)

FARE                1.000000
DISTANCE            0.670016
SW_No               0.543813
COUPON              0.496537
E_INCOME            0.326092
E_POP               0.285043
VACATION_No         0.276868
SLOT_Controlled     0.209438
S_INCOME            0.209135
GATE_Constrained    0.208540
S_POP               0.145097
NEW                 0.091730
HI                  0.025195
PAX                -0.090705
GATE_Free          -0.208540
SLOT_Free          -0.209438
VACATION_Yes       -0.276868
SW_Yes             -0.543813
Name: FARE, dtype: float64

In [188]:
fig = px.scatter(ADF, 
                 x="DISTANCE", 
                 y="FARE", 
                 height=500,
                 trendline='ols',
                 template='plotly_white')
fig.show()

# showing positively correlated

# 1. Create a simple predictive model of the target variable - "simple" meaning choose just ONE explanatory variable. 

In [189]:
X = ADF[['DISTANCE']].values
y = ADF[['FARE']].values

In [190]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=3)


In [191]:
linear_model1=LinearRegression()
linear_model1.fit(X_train, y_train)

LinearRegression()

In [192]:
from dmba import regressionSummary
# Evaluate Performance
# Training Performance
regressionSummary(y_train, linear_model1.predict(X_train))

# Test Performance
regressionSummary(y_test, linear_model1.predict(X_test))


Regression statistics

                      Mean Error (ME) : -0.0000
       Root Mean Squared Error (RMSE) : 55.8668
            Mean Absolute Error (MAE) : 46.0702
          Mean Percentage Error (MPE) : -14.9004
Mean Absolute Percentage Error (MAPE) : 35.3175

Regression statistics

                      Mean Error (ME) : -3.1413
       Root Mean Squared Error (RMSE) : 58.0845
            Mean Absolute Error (MAE) : 50.4340
          Mean Percentage Error (MPE) : -16.9390
Mean Absolute Percentage Error (MAPE) : 38.0777


In [193]:
#R-squared values for train & test data set.

y_pred_train = linear_model1.predict(X_train)
R2_train = r2_score(y_train, y_pred_train)

y_pred_test = linear_model1.predict(X_test)
R2_test = r2_score(y_test, y_pred_test)

print('R2 for training dataset',R2_train)
print('R2 for testing dataset',R2_test)

R2 for training dataset 0.46374225175587347
R2 for testing dataset 0.39902908644846047


In [194]:
X1b = ADF['DISTANCE']
y1b = ADF['FARE']

In [195]:
X_train, X_test, y_train, y_test = train_test_split(X1b, y1b, test_size=0.25, random_state=3)

In [196]:
X_train = sm.add_constant(X_train)

In [197]:
# for checking the p-values checking OLS summary
linear_model1b=sm.OLS(y_train,X_train).fit()
linear_model1b.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   FARE   R-squared:                       0.464
Model:                            OLS   Adj. R-squared:                  0.463
Method:                 Least Squares   F-statistic:                     411.6
Date:                Thu, 03 Nov 2022   Prob (F-statistic):           2.09e-66
Time:                        11:23:07   Log-Likelihood:                -2601.2
No. Observations:                 478   AIC:                             5206.
Df Residuals:                     476   BIC:                             5215.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         81.7647      4.632     17.652      0.000      72.663      90.866
DISTANCE       0.0819      0.004     20.289      0.000       0.074       0.090
==============================================================================
Omnibus:                       22.465   Durbin-Watson:                   2.307
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.155
Skew:                           0.531   Prob(JB):                     5.69e-06
Kurtosis:                       2.711   Cond. No.                     2.08e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.08e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [198]:
print('intercept:', linear_model1.intercept_) # Printing the interecept for linear regression
print('coefficient', linear_model1.coef_ )    # Printing the coefficient for linear regression


intercept: [81.76468303]
coefficient [[0.08189359]]


In [153]:
print ('Model Form Dataset 1(A)=>','FARE =', linear_model1.intercept_, '+', linear_model1.coef_,'* DISTANCE' )

Model Form Dataset 1(A)=> FARE = [81.76468303] + [[0.08189359]] * DISTANCE


# Does your model under or overfit the data?  How do you know?

No I don't the model overfits or underfits as there is not much diference in the MAPE value of Training and Testing datasets as we calculated above. I think model is working fine.

# Question 2

# Create a slightly more complicated predictive model of the target variable.  In particular, add 1-3 more variables that you think have potential to improve your model.

In [208]:
ADF = pd.read_csv(r'C:\Users\anagbhid\Documents\CourseWork_Q1\CIS508_Coursework\HandsOn3CIS_508\Airfares.csv')

ADF = ADF.drop(columns=[ 'S_CODE', 'S_CITY', 'E_CODE','E_CITY'])

ADF = pd.get_dummies(ADF, columns = ['SLOT', 'GATE', 'VACATION','SW'])

X1 = ADF.drop(columns=['FARE'])
y1 = ADF['FARE']



In [209]:
X_train, X_test, y_train, y_test = train_test_split(X1, y1, test_size=0.3, random_state=1)

In [210]:
from mlxtend.feature_selection import SequentialFeatureSelector as SFS

# selecting 5 best features - forward selection
Forward_Selection = SFS(linear_model1,
                        k_features=(1,5),
                        forward=True,
                        floating=False,
                        verbose=2,
                        scoring="r2",
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  11 out of  17 | elapsed:    0.8s remaining:    0.4s
[Parallel(n_jobs=-1)]: Done  17 out of  17 | elapsed:    0.8s finished

[2022-11-03 11:32:15] Features: 1/5 -- score: 0.4354576560486362[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  16 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  16 out of  16 | elapsed:    0.0s finished

[2022-11-03 11:32:15] Features: 2/5 -- score: 0.5880761592999174[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   8 out of  15 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  15 out of  15 | elapsed:    0.0s finished

[2022-11-03 11:32:15] Features: 3/5 -- score: 0.7004572411719286[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  14 | 

In [211]:
# selecting 5 best features - backward selection
Backward_Selection = SFS(linear_model1,
                        k_features=(1,6),
                        forward=False,
                        floating=False,
                        verbose=2,
                        scoring='r2',
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  11 out of  17 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  17 out of  17 | elapsed:    0.0s finished

[2022-11-03 11:32:16] Features: 16/1 -- score: 0.7579533855483624[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  10 out of  16 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  16 out of  16 | elapsed:    0.0s finished

[2022-11-03 11:32:16] Features: 15/1 -- score: 0.7581103256690737[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   8 out of  15 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  15 out of  15 | elapsed:    0.0s finished

[2022-11-03 11:32:16] Features: 14/1 -- score: 0.7582560096600284[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  14

In [212]:
print('Score for backward selected features',Backward_Selection.k_score_)
print()
print('Backard selected features',Backward_Selection.k_feature_names_)
print()
print('Score for forward selected features',Forward_Selection.k_score_)
print()
print('Forward selected features',Forward_Selection.k_feature_names_)

Score for backward selected features 0.7359054870157412

Backard selected features ('HI', 'S_POP', 'E_POP', 'DISTANCE', 'VACATION_No', 'SW_No')

Score for forward selected features 0.7289393215943825

Forward selected features ('HI', 'DISTANCE', 'SLOT_Controlled', 'VACATION_No', 'SW_No')


In [213]:
#selecting 3 best features - using EFS

from mlxtend.feature_selection import ExhaustiveFeatureSelector as EFS

Exhaustive_Selection = EFS(linear_model1,
                           min_features= 3,
                           max_features= 5,
                           scoring='r2',
                           n_jobs=-1).fit(X_train, y_train)


Features: 9248/9248

In [214]:
Exhaustive_Selection.best_score_

0.7289393215943826

In [215]:
Exhaustive_Selection.best_feature_names_

('HI', 'DISTANCE', 'SLOT_Free', 'VACATION_Yes', 'SW_No')

In [216]:
X2 = ADF[['DISTANCE','HI', 'VACATION_Yes','SLOT_Free', 'SW_No']].values
y2 = ADF['FARE'].values

In [217]:
X_train, X_test, y_train, y_test = train_test_split(X2, y2, test_size=0.25, random_state=3)


In [218]:
# considering following features to build the model - 'HI', 'DISTANCE', 'SW_No',

linear_model2=LinearRegression()
linear_model2.fit(X_train, y_train)

LinearRegression()

In [219]:
from dmba import regressionSummary
# Evaluate Performance
# Training Performance
regressionSummary(y_train, linear_model2.predict(X_train))

# Test Performance
regressionSummary(y_test, linear_model2.predict(X_test))


Regression statistics

                      Mean Error (ME) : -0.0000
       Root Mean Squared Error (RMSE) : 37.8169
            Mean Absolute Error (MAE) : 29.9863
          Mean Percentage Error (MPE) : -5.8641
Mean Absolute Percentage Error (MAPE) : 22.5196

Regression statistics

                      Mean Error (ME) : -3.0263
       Root Mean Squared Error (RMSE) : 39.8784
            Mean Absolute Error (MAE) : 31.3726
          Mean Percentage Error (MPE) : -7.9754
Mean Absolute Percentage Error (MAPE) : 24.2895


In [220]:
y_pred_train = linear_model2.predict(X_train)
R2_train = r2_score(y_train, y_pred_train)

y_pred_test = linear_model2.predict(X_test)
R2_test = r2_score(y_test, y_pred_test)

print('R2 for training dataset',R2_train)
print('R2 for testing dataset',R2_test)

R2 for training dataset 0.7542813895208544
R2 for testing dataset 0.7167250361030404


In [221]:
Complex_Model1a = smf.ols('FARE ~ DISTANCE + HI + VACATION_Yes ', data = ADF).fit()
Complex_Model1a.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   FARE   R-squared:                       0.607
Model:                            OLS   Adj. R-squared:                  0.605
Method:                 Least Squares   F-statistic:                     326.2
Date:                Thu, 03 Nov 2022   Prob (F-statistic):          4.68e-128
Time:                        11:38:18   Log-Likelihood:                -3370.2
No. Observations:                 638   AIC:                             6748.
Df Residuals:                     634   BIC:                             6766.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       44.7606      7.173      6.240      0.000      30.674      58.847
DISTANCE         0.0903      0.003     29.233      0.000       0.084       0.096
HI               0.0096      0.001      8.209      0.000       0.007       0.012
VACATION_Yes   -54.2033      4.334    -12.507      0.000     -62.714     -45.693
==============================================================================
Omnibus:                       17.098   Durbin-Watson:                   1.009
Prob(Omnibus):                  0.000   Jarque-Bera (JB):                8.841
Skew:                           0.011   Prob(JB):                       0.0120
Kurtosis:                       2.424   Cond. No.                     1.86e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.86e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [222]:
print('intercept:',linear_model2.intercept_) # Printing the interecept for linear regression
print('coefficient',linear_model2.coef_ )    # Printing the coefficient for linear regression

#'DISTANCE','HI', 'VACATION_Yes','SLOT_Free', 'SW_No'


intercept: 27.61369354246139
coefficient [ 8.11977535e-02  9.24927343e-03 -4.77141023e+01 -1.85699992e+01
  5.70476427e+01]


In [223]:
#Different model forms based on features.
print ('Model Form Dataset 1(A)=>','FARE =', linear_model1.intercept_, '+', linear_model1.coef_,'* DISTANCE' )
print()
print ('Model Form Dataset 1(B)=>','FARE =', linear_model2.intercept_, '+',8.11977535e-02 ,'* DISTANCE','+', 9.24927343e-03,'* HI', '-', 4.77141023e+01, '* VACATION_Yes', '-', 1.85699992e+01, '* SLOT_Free', '+', 5.70476427e+01, '* SW_No')

Model Form Dataset 1(A)=> FARE = [81.76468303] + [[0.08189359]] * DISTANCE

Model Form Dataset 1(B)=> FARE = 27.61369354246139 + 0.0811977535 * DISTANCE + 0.00924927343 * HI - 47.7141023 * VACATION_Yes - 18.5699992 * SLOT_Free + 57.0476427 * SW_No


# Take note of any differences in model performance from 1. to 2.


Comparing model 1 & 2, I can say that there is there is an increase in R-squared value, which I think is good. ALong with that we can see that MAPE value for Model 2 is reduced which is better, as less the error beter the model.

Performcance of model 2 is better.

# Do you notice any major changes in the magnitudes of your parameter estimates? 


Considering the feature "DISTANCE", we can see there is not much difference between the magnitudes of the parameter estimate. 

# Pick one parameter estimate and, in words, describe what it means?

Parameter estimates means, how a parameter affects or influences 1 unit change in the predictor contribution, so if we consider 'DISTANCE', it contributes or influences 0.0811 part of 1 unit of dependent variable or whatenver we are predicting in this case that is FARE. 

In simple terms if Fare increase Distance with have 0.0811 of contribution in the change.

In [224]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, RidgeCV,LassoCV
from sklearn.metrics import mean_absolute_percentage_error, mean_absolute_error, r2_score, mean_squared_error

# Question 3 - Applying LassoCV

In [225]:
ADF.columns

Index(['COUPON', 'NEW', 'HI', 'S_INCOME', 'E_INCOME', 'S_POP', 'E_POP',
       'DISTANCE', 'PAX', 'FARE', 'SLOT_Controlled', 'SLOT_Free',
       'GATE_Constrained', 'GATE_Free', 'VACATION_No', 'VACATION_Yes', 'SW_No',
       'SW_Yes'],
      dtype='object')

In [226]:
def train_test_metrics(X_train,X_test,y_train,y_test,model):
    names=['R2','MAE','MAPE','SSE'] 
    
    # Training Metrics
    y_hat = model.predict(X_train)
    
    # Create R2
    r2 = r2_score(y_train,y_hat).round(3)
    
    # Create MAE
    mae = mean_absolute_error(y_train,y_hat).round(3)
    
    # Create MAPE
    mape = mean_absolute_percentage_error(y_train,y_hat).round(3)*100
    
    # Create MSE
    mse = (mean_squared_error(y_train,y_hat)*len(y_train)).round(3)
    
    train_metrics = [r2, mae, mape,mse]
    train_metrics = pd.DataFrame({'Train':train_metrics},index=names)
    
    # Testing Metrics
    y_hat = model.predict(X_test)
    test_metrics = [r2_score(y_test,y_hat).round(3),
                    mean_absolute_error(y_test,y_hat).round(3),
                    mean_absolute_percentage_error(y_test,y_hat).round(3)*100,
                    (mean_squared_error(y_test,y_hat)*len(y_test)).round(3)
                   ]
    test_metrics = pd.DataFrame({'Test':test_metrics},index=names)
    
    all_metrics = train_metrics.merge(test_metrics,left_index=True,right_index=True)
    
    print(pd.DataFrame({'Predictor':X_train.columns, 'coefficent':model.coef_.round(3)}))
    print('\n')
    print(all_metrics)
    print('\n')

In [227]:
predictors = ['COUPON', 'NEW', 'HI', 'S_INCOME', 'E_INCOME', 'S_POP', 'E_POP',
       'DISTANCE', 'PAX', 'SLOT_Controlled', 'SLOT_Free',
       'GATE_Constrained', 'GATE_Free', 'VACATION_No', 'VACATION_Yes', 'SW_No',
       'SW_Yes'
             ]
X = ADF[predictors]
y = ADF['FARE']
Xtrain,Xtest,ytrain,ytest = train_test_split(X,y,test_size=.25,random_state=1)

r_alphas = [0.04,0.06,0.08,0.085,0.09,0.099,.1,0.11,0.2,0.25]

model = LassoCV(alphas=r_alphas,normalize=True)
model.fit(Xtrain,ytrain)

train_test_metrics(Xtrain,Xtest,ytrain,ytest,model)

           Predictor  coefficent
0             COUPON       0.000
1                NEW      -1.232
2                 HI       0.007
3           S_INCOME       0.001
4           E_INCOME       0.001
5              S_POP       0.000
6              E_POP       0.000
7           DISTANCE       0.075
8                PAX      -0.001
9    SLOT_Controlled      14.990
10         SLOT_Free      -0.000
11  GATE_Constrained      19.569
12         GATE_Free      -0.000
13       VACATION_No      35.904
14      VACATION_Yes      -0.000
15             SW_No      40.043
16            SW_Yes      -0.000


           Train        Test
R2         0.781       0.794
MAE       27.459      28.062
MAPE      20.500      20.800
SSE   595428.012  198752.624




C:\Users\anagbhid\Anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Lasso())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * np.sqrt(n_samples). 



In [228]:
print('LassoCV Alpha Selection: ',model.alpha_)

LassoCV Alpha Selection:  0.04


In [229]:
print('intercept for lasso', model.intercept_) # Printing the interecept for linear regression
print('coefficients', model.coef_ )   # Printing the coefficients for linear regression

intercept for lasso -82.20940279580753
coefficients [ 0.00000000e+00 -1.23236746e+00  7.27934817e-03  7.95671504e-04
  1.45894034e-03  3.06047256e-06  3.42093731e-06  7.49706955e-02
 -6.96035868e-04  1.49902829e+01 -3.27108365e-15  1.95691100e+01
 -1.82523746e-14  3.59036910e+01 -0.00000000e+00  4.00428351e+01
 -1.77088224e-15]


In [230]:
print ('Model Form 1=>','FARE =', linear_model1.intercept_, '+', linear_model1.coef_,'* DISTANCE' )
print()
print ('Model Form 2=>','FARE =', linear_model2.intercept_, '+',8.11977535e-02 ,'* DISTANCE','+', 9.24927343e-03,'* HI', '-', 4.77141023e+01, '* VACATION_Yes', '-', 1.85699992e+01, '* SLOT_Free', '+', 5.70476427e+01, '* SW_No')
print()
print ('Model Form 3=>','FARE =', model.intercept_, '-',1.232 ,'* NEW','+', 0.075,'* DISTANCE', '-',14.990, '* SLOT_Controlled', '-', 19.569, '* GATE_Constrained', '+', 35.904,'* VACATION_No', '+', 40.043, '* SW_No')

print()
print("Mean Absolute Percentage Error (MAPE) for model 1 : 38.0777")
print("Mean Absolute Percentage Error (MAPE) for model 2 : 24.2895")
print("Mean Absolute Percentage Error (MAPE) for model 3 :  20.800")

Model Form 1=> FARE = [81.76468303] + [[0.08189359]] * DISTANCE

Model Form 2=> FARE = 27.61369354246139 + 0.0811977535 * DISTANCE + 0.00924927343 * HI - 47.7141023 * VACATION_Yes - 18.5699992 * SLOT_Free + 57.0476427 * SW_No

Model Form 3=> FARE = -82.20940279580753 - 1.232 * NEW + 0.075 * DISTANCE - 14.99 * SLOT_Controlled - 19.569 * GATE_Constrained + 35.904 * VACATION_No + 40.043 * SW_No

Mean Absolute Percentage Error (MAPE) for model 1 : 38.0777
Mean Absolute Percentage Error (MAPE) for model 2 : 24.2895
Mean Absolute Percentage Error (MAPE) for model 3 :  20.800


# DATASET 2 - Boston Housing

In [231]:
BDF=pd.read_csv(r'C:\Users\anagbhid\Documents\CourseWork_Q1\CIS508_Coursework\HandsOn3CIS_508\BostonHousing.csv')
# BDF.columns= BDF.columns.str.lower()
BDF.columns = BDF.columns.str.replace('. ', '')
BDF

C:\Users\anagbhid\AppData\Local\Temp\ipykernel_30508\1299147145.py:3: FutureWarning:

The default value of regex will change from True to False in a future version.



CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO  LSTAT  MEDV  CATMEDV  
0       15.3   4.98  24.0        0  
1       17.8   9.14  21.6        0  
2       17.8   4.03  34.7        1  
3       18.7   2.94  33.4        1  
4       18.7   5.33  36.2        1  
..       ...    ...   ...      ...  
501     21.0   9.67  22.4        0  
502     21.0   9.08  20.6        0  
503     21.0   5.64  23.9        0  
504     21.0   6.48  22.0        0  
505     21.0   7.88  11.9        0  

[506 rows x 14 columns]

In [232]:
BDFcorr_matrix = BDF.corr()
BDFcorr_matrix["CATMEDV"].sort_values(ascending=False)

CATMEDV    1.000000
MEDV       0.789789
RM         0.641265
ZN         0.365296
DIS        0.118887
CHAS       0.108631
CRIM      -0.151987
AGE       -0.191196
RAD       -0.197924
NOX       -0.232502
TAX       -0.273687
INDUS     -0.366276
PTRATIO   -0.443425
LSTAT     -0.469911
Name: CATMEDV, dtype: float64

In [233]:
fig = px.scatter(BDF, 
                 x="LSTAT", 
                 y="MEDV", 
                 height=500,
                 trendline='ols',
                 template='plotly_white')
fig.show()

In [234]:
X2a = BDF[['LSTAT']].values
y2a = BDF['MEDV'].values

In [235]:
X_train, X_test, y_train, y_test = train_test_split(X2a, y2a, test_size=0.25, random_state=3)


In [236]:
linear_model3=LinearRegression()
linear_model3.fit(X_train, y_train)

LinearRegression()

In [237]:
# Evaluate Performance
# Training Performance
regressionSummary(y_train, linear_model3.predict(X_train))

# Test Performance
regressionSummary(y_test, linear_model3.predict(X_test))


Regression statistics

                      Mean Error (ME) : -0.0000
       Root Mean Squared Error (RMSE) : 6.2158
            Mean Absolute Error (MAE) : 4.5483
          Mean Percentage Error (MPE) : -5.7269
Mean Absolute Percentage Error (MAPE) : 22.1874

Regression statistics

                      Mean Error (ME) : -0.2139
       Root Mean Squared Error (RMSE) : 6.1676
            Mean Absolute Error (MAE) : 4.4298
          Mean Percentage Error (MPE) : -5.8581
Mean Absolute Percentage Error (MAPE) : 19.2462


In [238]:
y_pred_train = linear_model3.predict(X_train)
R2_train = r2_score(y_train, y_pred_train)

y_pred_test = linear_model3.predict(X_test)
R2_test = r2_score(y_test, y_pred_test)

print('R2 for training dataset',R2_train)
print('R2 for testing dataset',R2_test)

R2 for training dataset 0.5483569257040284
R2 for testing dataset 0.5307275781842369


In [239]:
X2aa = BDF['LSTAT']
y2aa = BDF['MEDV']

In [240]:
X_train, X_test, y_train, y_test = train_test_split(X2aa, y2aa, test_size=0.25, random_state=3)

In [241]:
X_train = sm.add_constant(X_train)

In [242]:
# for checking the p-values checking OLS summary
linear_model2aa=sm.OLS(y_train,X_train).fit()
linear_model2aa.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.548
Model:                            OLS   Adj. R-squared:                  0.547
Method:                 Least Squares   F-statistic:                     457.7
Date:                Thu, 03 Nov 2022   Prob (F-statistic):           4.70e-67
Time:                        11:47:10   Log-Likelihood:                -1230.2
No. Observations:                 379   AIC:                             2464.
Df Residuals:                     377   BIC:                             2472.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         34.5786      0.648     53.365      0.000      33.304      35.853
LSTAT         -0.9478      0.044    -21.395      0.000      -1.035      -0.861
==============================================================================
Omnibus:                      104.605   Durbin-Watson:                   1.819
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              219.803
Skew:                           1.444   Prob(JB):                     1.86e-48
Kurtosis:                       5.362   Cond. No.                         29.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [243]:
print('intercept:', linear_model3.intercept_) # Printing the interecept for linear regression)
print('coefficient', linear_model3.coef_ ) # Printing the coefficient for linear regression)

intercept: 34.57857704438267
coefficient [-0.94776211]


In [259]:
print ('Model Form 1 =>','MEDV =', linear_model3.intercept_, '+', linear_model3.coef_, '* LSTAT')

Model Form 1 => MEDV = 34.57857704438267 + [-0.94776211] * LSTAT


# Does your model under or overfit the data?  How do you know?

No I don't the model overfits or underfits as there is not much diference in the MAPE value of Training and
Testing datasets as we calculated above. I think model is working fine.

# Question 2

# Create a slightly more complicated predictive model of the target variable.  In particular, add 1-3 more variables that you think have potential to improve your model.

In [245]:
BDF=pd.read_csv(r'C:\Users\anagbhid\Documents\CourseWork_Q1\CIS508_Coursework\HandsOn3CIS_508\BostonHousing.csv')

X2b = BDF.drop(columns=['MEDV'])
y2b = BDF['MEDV']

In [246]:
X_train, X_test, y_train, y_test = train_test_split(X2b, y2b, test_size=0.3, random_state=1)

In [247]:
Forward_Selection = SFS(linear_model3,
                        k_features=(1,5),
                        forward=True,
                        floating=False,
                        verbose=2,
                        scoring="r2",
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of  13 | elapsed:    3.4s remaining:    5.4s
[Parallel(n_jobs=-1)]: Done  13 out of  13 | elapsed:    3.4s finished

[2022-11-03 11:48:25] Features: 1/5 -- score: 0.6012431653338002[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   4 out of  12 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  12 out of  12 | elapsed:    0.0s finished

[2022-11-03 11:48:25] Features: 2/5 -- score: 0.7750075455445586[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   2 out of  11 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   8 out of  11 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  11 out of  11 | elapsed:    0.0s finished

[2022-11-03 11:48:25] Features: 3/5 -- score: 0.7842855238090033[Parallel(n_jobs=-1)]: Using backend Lok

In [248]:
Backward_Selection = SFS(linear_model3,
                        k_features=(1,6),
                        forward=False,
                        floating=False,
                        verbose=2,
                        scoring='r2',
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of  13 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  13 out of  13 | elapsed:    0.0s finished

[2022-11-03 11:48:25] Features: 12/1 -- score: 0.8136321373418337[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   4 out of  12 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  12 out of  12 | elapsed:    0.0s finished

[2022-11-03 11:48:25] Features: 11/1 -- score: 0.8152317292377482[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   2 out of  11 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   8 out of  11 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  11 out of  11 | elapsed:    0.0s finished

[2022-11-03 11:48:25] Features: 10/1 -- score: 0.8155785109165967[Parallel(n_jobs=-1)]: Using backend 

In [249]:
print('Score for backward selected features',Backward_Selection.k_score_)
print()
print('Backard selected features',Backward_Selection.k_feature_names_)
print()
print('Score for forward selected features',Forward_Selection.k_score_)
print()
print('Forward selected features',Forward_Selection.k_feature_names_)

Score for backward selected features 0.8028580340869785

Backard selected features ('CHAS', 'NOX', 'DIS', 'PTRATIO', 'LSTAT', 'CAT. MEDV')

Score for forward selected features 0.7955164034189278

Forward selected features ('CRIM', 'CHAS', 'DIS', 'LSTAT', 'CAT. MEDV')


In [251]:
# from mlxtend.feature_selection import ExhaustiveFeatureSelector as EFS

Exhaustive_Selection = EFS(linear_model3,
                           min_features= 3,
                           max_features= 5,
                           scoring='r2',
                           n_jobs=-1).fit(X_train, y_train)

Features: 2288/2288

In [ ]:
Exhaustive_Selection.best_score_

In [ ]:

Exhaustive_Selection.best_feature_names_


In [252]:
X2c = BDF[['CHAS', 'DIS', 'LSTAT']].values
y2c = BDF['MEDV'].values

In [253]:
X_train, X_test, y_train, y_test = train_test_split(X2c, y2c, test_size=0.25, random_state=3)

In [254]:
linear_model4=LinearRegression()
linear_model4.fit(X_train, y_train)

LinearRegression()

In [255]:
# Evaluate Performance
# Training Performance
regressionSummary(y_train, linear_model4.predict(X_train))

# Test Performance
regressionSummary(y_test, linear_model4.predict(X_test))


Regression statistics

                      Mean Error (ME) : 0.0000
       Root Mean Squared Error (RMSE) : 5.9710
            Mean Absolute Error (MAE) : 4.4046
          Mean Percentage Error (MPE) : -5.5054
Mean Absolute Percentage Error (MAPE) : 21.9328

Regression statistics

                      Mean Error (ME) : -0.1964
       Root Mean Squared Error (RMSE) : 6.0441
            Mean Absolute Error (MAE) : 4.3228
          Mean Percentage Error (MPE) : -5.9882
Mean Absolute Percentage Error (MAPE) : 18.8359


In [256]:
y_pred_train = linear_model4.predict(X_train)
R2_train = r2_score(y_train, y_pred_train)

y_pred_test = linear_model4.predict(X_test)
R2_test = r2_score(y_test, y_pred_test)

print('R2 for training dataset',R2_train)
print('R2 for testing dataset',R2_test)

R2 for training dataset 0.5832265220350934
R2 for testing dataset 0.5493349750990879


In [257]:
Complex_Model2ab = smf.ols('MEDV ~ CHAS + DIS + LSTAT', data = BDF).fit()
Complex_Model2ab.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.576
Model:                            OLS   Adj. R-squared:                  0.573
Method:                 Least Squares   F-statistic:                     227.3
Date:                Thu, 03 Nov 2022   Prob (F-statistic):           3.99e-93
Time:                        11:48:53   Log-Likelihood:                -1623.2
No. Observations:                 506   AIC:                             3254.
Df Residuals:                     502   BIC:                             3271.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     37.4851      1.014     36.983      0.000      35.494      39.476
CHAS           4.3020      1.065      4.039      0.000       2.209       6.395
DIS           -0.5900      0.148     -3.991      0.000      -0.881      -0.300
LSTAT         -1.0283      0.043    -23.668      0.000      -1.114      -0.943
==============================================================================
Omnibus:                      111.240   Durbin-Watson:                   1.031
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              205.531
Skew:                           1.253   Prob(JB):                     2.34e-45
Kurtosis:                       4.861   Cond. No.                         64.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [258]:
print('intercept = ', linear_model4.intercept_) # Printing the interecept for linear regression
print('coefficients', linear_model4.coef_ )   # Printing the coefficients for linear regression


intercept =  36.69290310118993
coefficients [ 5.10748966 -0.49192686 -0.99558541]


In [260]:
print ('Model Form 2=>','MEDV =', linear_model4.intercept_, '+',5.10748966 ,'* CHAS','-', 0.49192686,'* DIS', '-', 0.99558541, '* LSTAT')

Model Form 2=> MEDV = 36.69290310118993 + 5.10748966 * CHAS - 0.49192686 * DIS - 0.99558541 * LSTAT


# Take note of any differences in model performance from 1. to 2.


MAPE for Model 2 is less than MAPE for model 1, Model 2 performs better than one. 

# Do you notice any major changes in the magnitudes of your parameter estimates?  


Considering LSTAT feature there is not much difference in the change in the magnitudes, for model 1-0.94776211 & model 2 -0.99558541

# Pick one parameter estimate and, in words, describe what it means?

Considering LSTAT here it is negatively correlated to the MEDV value, so LSTAT influences inversly on the MEDV values, as LSTAT increses MEDV value decreases.

# Question 3 

In [271]:
BDF.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'LSTAT', 'MEDV', 'CAT. MEDV'],
      dtype='object')

In [272]:
BDF.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   LSTAT  MEDV  CAT. MEDV  
0   4.98  24.0          0  
1   9.14  21.6          0  
2   4.03  34.7          1  
3   2.94  33.4          1  
4   5.33  36.2          1

In [330]:
# considering only necessary variables.
predictors = ['CRIM', 'ZN', 'INDUS', 'NOX', 'RM', 'AGE', 'DIS', 'TAX',
       'PTRATIO', 'LSTAT'
             ]
X = BDF[predictors]
y = BDF['MEDV']

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=.25,random_state=1)

r_alphas = [0.04,0.06,0.08,0.085,0.09,0.099,.1,0.11,0.2,0.25]

model = LassoCV(alphas=r_alphas,normalize=True)
model.fit(X_train,y_train)

train_test_metrics(X_train,X_test,y_train,y_test,model2l)

  Predictor  coefficent
0      CRIM      -0.009
1        ZN       0.000
2     INDUS      -0.000
3       NOX      -0.000
4        RM       3.058
5       AGE      -0.000
6       DIS      -0.000
7       TAX      -0.000
8   PTRATIO      -0.695
9     LSTAT      -0.545


          Train      Test
R2        0.644     0.684
MAE       3.627     4.254
MAPE     18.500    20.500
SSE   10711.928  3974.969




C:\Users\anagbhid\Anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Lasso())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * np.sqrt(n_samples). 



In [331]:
print('LassoCV Alpha Selection: ',model.alpha_)

LassoCV Alpha Selection:  0.04


In [275]:
print ('Model Form 1=>','MEDV =', linear_model3.intercept_, '+', linear_model3.coef_,'* LSTAT' )
print()
print ('Model Form 2=>','FARE =', linear_model4.intercept_, '+', 5.10748966 ,'* CHAS','-', 0.49192686,'* DIS', '-', 0.99558541, '* LSTAT',)
print()
print ('Model Form 3=>','FARE =', model.intercept_, '-',3.058 ,'* RM','+', 0.075,'* DISTANCE', '-',0.695, '* PTRATIO', '-', 0.545, '* LSTAT')

print()
print("Mean Absolute Percentage Error (MAPE) for model 1 : 19.2462")
print("Mean Absolute Percentage Error (MAPE) for model 2 : 18.8359")
print("Mean Absolute Percentage Error (MAPE) for model 3 :  20.500")

Model Form 1=> MEDV = 34.57857704438267 + [-0.94776211] * LSTAT

Model Form 2=> FARE = 36.69290310118993 + 5.10748966 * CHAS - 0.49192686 * DIS - 0.99558541 * LSTAT

Model Form 3=> FARE = 23.13832084677211 - 3.058 * RM + 0.075 * DISTANCE - 0.695 * PTRATIO - 0.545 * LSTAT

Mean Absolute Percentage Error (MAPE) for model 1 : 19.2462
Mean Absolute Percentage Error (MAPE) for model 2 : 18.8359
Mean Absolute Percentage Error (MAPE) for model 3 :  20.500


# DATASET 3 - Toyota Dataset

In [276]:
TDF=pd.read_csv(r'C:\Users\anagbhid\Documents\CourseWork_Q1\CIS508_Coursework\HandsOn3CIS_508\ToyotaCorolla.csv')
TDF = TDF.drop(columns=['Model'])
TDF = pd.get_dummies(TDF, columns = ['Fuel_Type', 'Color'])
TDF

Id  Price  Age_08_04  Mfg_Month  Mfg_Year     KM   HP  Met_Color  \
0        1  13500         23         10      2002  46986   90          1   
1        2  13750         23         10      2002  72937   90          1   
2        3  13950         24          9      2002  41711   90          1   
3        4  14950         26          7      2002  48000   90          0   
4        5  13750         30          3      2002  38500   90          0   
...    ...    ...        ...        ...       ...    ...  ...        ...   
1431  1438   7500         69         12      1998  20544   86          1   
1432  1439  10845         72          9      1998  19000   86          0   
1433  1440   8500         71         10      1998  17016   86          0   
1434  1441   7250         70         11      1998  16916   86          1   
1435  1442   6950         76          5      1998      1  110          0   

      Automatic    CC  ...  Color_Beige  Color_Black  Color_Blue  Color_Green  \
0             0  2000  ...            0            0           1            0   
1             0  2000  ...            0            0           0            0   
2             0  2000  ...            0            0           1            0   
3             0  2000  ...            0            1           0            0   
4             0  2000  ...            0            1           0            0   
...         ...   ...  ...          ...          ...         ...          ...   
1431          0  1300  ...            0            0           1            0   
1432          0  1300  ...            0            0           0            0   
1433          0  1300  ...            0            0           1            0   
1434          0  1300  ...            0            0           0            0   
1435          0  1600  ...            0            0           0            1   

      Color_Grey  Color_Red  Color_Silver  Color_Violet  Color_White  \
0              0          0             0             0            0   
1              0          0             1             0            0   
2              0          0             0             0            0   
3              0          0             0             0            0   
4              0          0             0             0            0   
...          ...        ...           ...           ...          ...   
1431           0          0             0             0            0   
1432           1          0             0             0            0   
1433           0          0             0             0            0   
1434           1          0             0             0            0   
1435           0          0             0             0            0   

      Color_Yellow  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
1431             0  
1432             0  
1433             0  
1434             0  
1435             0  

[1436 rows x 49 columns]

In [277]:
#Choosing the variable with highest correlation - "Age_08_04"
TDFcorr_matrix = TDF.corr()
TDFcorr_matrix["Price"].sort_values(ascending=False)

Price                1.000000
Mfg_Year             0.885159
Boardcomputer        0.601292
Automatic_airco      0.588262
Weight               0.581198
CD_Player            0.481374
Airco                0.429259
Powered_Windows      0.356518
Central_Lock         0.343458
HP                   0.314990
ABS                  0.306138
Airbag_2             0.248974
Mistlamps            0.222083
Quarterly_Tax        0.219197
Mfr_Guarantee        0.197802
Doors                0.185326
Color_Grey           0.169947
Sport_Model          0.164121
Guarantee_Period     0.146627
CC                   0.126389
Met_Color            0.108905
Metallic_Rim         0.108564
Backseat_Divider     0.102569
Airbag_1             0.093588
Power_Steering       0.064275
Gears                0.063104
Fuel_Type_Diesel     0.054084
Parking_Assistant    0.044375
Color_Black          0.034896
Automatic            0.033081
Color_Silver         0.028562
BOVAG_Guarantee      0.028133
Color_Yellow         0.022726
Color_Blue

In [280]:
fig = px.scatter(TDF, 
                 x="Age_08_04", 
                 y="Price", 
                 height=500,
                 trendline='ols',
                 template='plotly_white')
fig.show()

# showing negatively correlated

In [281]:
# from sklearn.preprocessing import MinMaxScaler
# scaler = MinMaxScaler()
# TDF_norm = pd.DataFrame(scaler.fit_transform(TDF), columns=TDF.columns)
# TDF_norm

In [282]:
X3a = TDF[['Age_08_04']].values
y3a = TDF['Price'].values


In [283]:
X_train, X_test, y_train, y_test = train_test_split(X3a, y3a, test_size=0.3, random_state=3)

In [284]:
linear_model3a=LinearRegression()
linear_model3a.fit(X_train, y_train)

LinearRegression()

In [285]:
from dmba import regressionSummary
# Evaluate Performance
# Training Performance
regressionSummary(y_train, linear_model3a.predict(X_train))

# Test Performance
regressionSummary(y_test, linear_model3a.predict(X_test))


Regression statistics

                      Mean Error (ME) : -0.0000
       Root Mean Squared Error (RMSE) : 1788.5342
            Mean Absolute Error (MAE) : 1248.2930
          Mean Percentage Error (MPE) : -2.0062
Mean Absolute Percentage Error (MAPE) : 12.1768

Regression statistics

                      Mean Error (ME) : -178.4958
       Root Mean Squared Error (RMSE) : 1641.5998
            Mean Absolute Error (MAE) : 1257.1250
          Mean Percentage Error (MPE) : -3.4237
Mean Absolute Percentage Error (MAPE) : 12.2215


In [286]:
#R-squared values for train & test data set.

y_pred_train = linear_model3a.predict(X_train)
R2_train = r2_score(y_train, y_pred_train)

y_pred_test = linear_model3a.predict(X_test)
R2_test = r2_score(y_test, y_pred_test)

print('R2 for training dataset',R2_train)
print('R2 for testing dataset',R2_test)

R2 for training dataset 0.7546807884728556
R2 for testing dataset 0.7976711887725145


In [287]:
X3aa = TDF['Age_08_04']
y3aa = TDF['Price']


In [288]:
X_train, X_test, y_train, y_test = train_test_split(X3aa, y3aa, test_size=0.25, random_state=3)

In [289]:
X_train = sm.add_constant(X_train)

In [290]:
# for checking the p-values checking OLS summary
linear_model3aa=sm.OLS(y_train,X_train).fit()
linear_model3aa.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.766
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     3516.
Date:                Thu, 03 Nov 2022   Prob (F-statistic):               0.00
Time:                        12:57:54   Log-Likelihood:                -9601.5
No. Observations:                1077   AIC:                         1.921e+04
Df Residuals:                    1075   BIC:                         1.922e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       2.055e+04    174.599    117.722      0.000    2.02e+04    2.09e+04
Age_08_04   -174.4404      2.942    -59.296      0.000    -180.213    -168.668
==============================================================================
Omnibus:                      293.592   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2269.570
Skew:                           1.033   Prob(JB):                         0.00
Kurtosis:                       9.805   Cond. No.                         189.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [314]:
print('intercept:', linear_model3a.intercept_) # Printing the interecept for linear regression)
print('coefficient', linear_model3a.coef_ ) # Printing the coefficient for linear regression)



intercept: 20400.557220502684
coefficient [-171.87956797]


In [316]:

print ('Model Form 1 =>','Price =', linear_model3a.intercept_, '+', linear_model3a.coef_, '* Age_08_04')

Model Form 1 => Price = 20400.557220502684 + [-171.87956797] * Age_08_04


# Does your model under or overfit the data?  How do you know?

Comparing the MAPE values for testing and training I can see that the model is working fine, i don't thinks it is overfiitting or underfitting.

# Question 2

# Create a slightly more complicated predictive model of the target variable.  In particular, add 1-3 more variables that you think have potential to improve your model.

In [291]:
X3b = TDF.drop(columns=['Price'])
y3b = TDF['Price']

In [292]:
X_train, X_test, y_train, y_test = train_test_split(X3b, y3b, test_size=0.3, random_state=3)

In [293]:
Forward_Selection = SFS(linear_model3a,
                        k_features=(1,5),
                        forward=True,
                        floating=False,
                        verbose=2,
                        scoring="r2",
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  25 tasks      | elapsed:    3.1s
[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    3.2s finished

[2022-11-03 12:59:33] Features: 1/5 -- score: 0.7691140193704991[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  32 out of  47 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  47 out of  47 | elapsed:    0.0s finished

[2022-11-03 12:59:33] Features: 2/5 -- score: 0.8232402914150987[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  31 out of  46 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  46 out of  46 | elapsed:    0.0s finished

[2022-11-03 12:59:33] Features: 3/5 -- score: 0.8454672054703453[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  30 out of  45 | elapsed:    0.0s re

In [294]:
Backward_Selection = SFS(linear_model3a,
                        k_features=(1,6),
                        forward=False,
                        floating=False,
                        verbose=2,
                        scoring='r2',
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    0.1s finished

[2022-11-03 12:59:33] Features: 47/1 -- score: 0.9028779838983935[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  32 out of  47 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  47 out of  47 | elapsed:    0.1s finished

[2022-11-03 12:59:34] Features: 46/1 -- score: 0.9036448109641363[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  31 out of  46 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  46 out of  46 | elapsed:    0.1s finished

[2022-11-03 12:59:34] Features: 45/1 -- score: 0.9041446362626873[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  30 out of  45 | elapsed:    0.1s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  45 out of  45

[Parallel(n_jobs=-1)]: Done  10 out of  16 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  16 out of  16 | elapsed:    0.0s finished

[2022-11-03 12:59:37] Features: 15/1 -- score: 0.9042006750268994[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   8 out of  15 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  15 out of  15 | elapsed:    0.0s finished

[2022-11-03 12:59:37] Features: 14/1 -- score: 0.903459707516784[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   7 out of  14 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  14 out of  14 | elapsed:    0.0s finished

[2022-11-03 12:59:37] Features: 13/1 -- score: 0.9027490487572148[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   5 out of  13 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  13 out of

In [295]:
print('Score for backward selected features',Backward_Selection.k_score_)
print()
print('Backard selected features',Backward_Selection.k_feature_names_)
print()
print('Score for forward selected features',Forward_Selection.k_score_)
print()
print('Forward selected features',Forward_Selection.k_feature_names_)

Score for backward selected features 0.8826614634555601

Backard selected features ('Mfg_Year', 'KM', 'HP', 'CC', 'Weight', 'Automatic_airco')

Score for forward selected features 0.880655289736217

Forward selected features ('Id', 'Mfg_Year', 'HP', 'Weight', 'Automatic_airco')


In [296]:
TDF.columns

Index(['Id', 'Price', 'Age_08_04', 'Mfg_Month', 'Mfg_Year', 'KM', 'HP',
       'Met_Color', 'Automatic', 'CC', 'Doors', 'Cylinders', 'Gears',
       'Quarterly_Tax', 'Weight', 'Mfr_Guarantee', 'BOVAG_Guarantee',
       'Guarantee_Period', 'ABS', 'Airbag_1', 'Airbag_2', 'Airco',
       'Automatic_airco', 'Boardcomputer', 'CD_Player', 'Central_Lock',
       'Powered_Windows', 'Power_Steering', 'Radio', 'Mistlamps',
       'Sport_Model', 'Backseat_Divider', 'Metallic_Rim', 'Radio_cassette',
       'Parking_Assistant', 'Tow_Bar', 'Fuel_Type_CNG', 'Fuel_Type_Diesel',
       'Fuel_Type_Petrol', 'Color_Beige', 'Color_Black', 'Color_Blue',
       'Color_Green', 'Color_Grey', 'Color_Red', 'Color_Silver',
       'Color_Violet', 'Color_White', 'Color_Yellow'],
      dtype='object')

In [297]:
#removing columns with Yes-No
TDF = pd.read_csv(r'C:\Users\anagbhid\Documents\CourseWork_Q1\CIS508_Coursework\HandsOn3CIS_508\ToyotaCorolla.csv')
TDF = TDF.drop(columns=['Model'])
TDF = pd.get_dummies(TDF, columns = ['Fuel_Type', 'Color'])
TDF = TDF.drop(columns=['Id', 'Mfg_Month', 'Mfg_Year','Quarterly_Tax','Automatic_airco', 'Automatic_airco','ABS', 'Airbag_1', 'Airbag_2','Boardcomputer', 'CD_Player', 'Central_Lock',
       'Powered_Windows', 'Power_Steering', 'Radio', 'Mistlamps',
       'Sport_Model', 'Backseat_Divider', 'Metallic_Rim', 'Radio_cassette',
       'Parking_Assistant', 'Tow_Bar'])

In [298]:
X3b = TDF.drop(columns=['Price'])
y3b = TDF['Price']

In [299]:
X_train, X_test, y_train, y_test = train_test_split(X3b, y3b, test_size=0.3, random_state=3)

In [300]:
Forward_Selection = SFS(linear_model3a,
                        k_features=(1,5),
                        forward=True,
                        floating=False,
                        verbose=2,
                        scoring="r2",
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  27 out of  27 | elapsed:    0.0s finished

[2022-11-03 12:59:51] Features: 1/5 -- score: 0.7524693730736042[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  26 out of  26 | elapsed:    0.0s finished

[2022-11-03 12:59:51] Features: 2/5 -- score: 0.7953942981084413[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  23 out of  25 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  25 out of  25 | elapsed:    0.0s finished

[2022-11-03 12:59:51] Features: 3/5 -- score: 0.8378650735737473[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  22 out of  24 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  24 out of  24 | elapsed:    0.0s finished

[2022-11-03 12:59:51] Features: 4/5 -- score: 0.855734

In [301]:
Backward_Selection = SFS(linear_model3a,
                        k_features=(1,6),
                        forward=False,
                        floating=False,
                        verbose=2,
                        scoring='r2',
                        n_jobs=-1,
                       cv=5).fit(X_train, y_train)

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  27 out of  27 | elapsed:    0.0s finished

[2022-11-03 12:59:53] Features: 26/1 -- score: 0.8700931864131791[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  26 out of  26 | elapsed:    0.0s finished

[2022-11-03 12:59:53] Features: 25/1 -- score: 0.8706369279636853[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  23 out of  25 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  25 out of  25 | elapsed:    0.0s finished

[2022-11-03 12:59:53] Features: 24/1 -- score: 0.8709496098316942[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  22 out of  24 | elapsed:    0.0s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done  24 out of  24 | elapsed:    0.0s finished

[2022-11-03 12:59:53] Features: 23/1 -- score: 0.87

In [ ]:
print('Score for backward selected features',Backward_Selection.k_score_)
print()
print('Backard selected features',Backward_Selection.k_feature_names_)
print()
print('Score for forward selected features',Forward_Selection.k_score_)
print()
print('Forward selected features',Forward_Selection.k_feature_names_)

In [ ]:
# from mlxtend.feature_selection import ExhaustiveFeatureSelector as EFS

# Exhaustive_Selection = EFS(linear_model3a,
#                            min_features= 3,
#                            max_features= 5,
#                            scoring='r2',
#                            n_jobs=-1).fit(X_train, y_train)

In [ ]:
# Exhaustive_Selection.best_score_

In [302]:
# Exhaustive_Selection.best_feature_names_

('CRIM', 'CHAS', 'DIS', 'LSTAT', 'CAT. MEDV')

In [303]:
pd.DataFrame.from_dict(Forward_Selection.get_metric_dict()).T

feature_idx                                          cv_scores  \
1             (0,)  [0.7349323689326035, 0.761324938607378, 0.7043...   
2           (0, 9)  [0.7956436929415582, 0.787270788055537, 0.7456...   
3        (0, 1, 9)  [0.8552431427246281, 0.821287671646128, 0.7986...   
4     (0, 1, 2, 9)  [0.8591903620338608, 0.8306046098225779, 0.840...   
5  (0, 1, 2, 5, 9)  [0.8685218034303288, 0.8098226298380534, 0.849...   

  avg_score                    feature_names  ci_bound   std_dev   std_err  
1  0.752469                     (Age_08_04,)  0.038216  0.029734  0.014867  
2  0.795394              (Age_08_04, Weight)  0.037301  0.029022  0.014511  
3  0.837865          (Age_08_04, KM, Weight)  0.031648  0.024623  0.012312  
4  0.855734      (Age_08_04, KM, HP, Weight)  0.023517  0.018297  0.009149  
5  0.858812  (Age_08_04, KM, HP, CC, Weight)  0.035465  0.027593  0.013797

In [304]:
pd.DataFrame.from_dict(Backward_Selection.get_metric_dict()).T

feature_idx  \
27  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...   
26  (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12, 13, 14...   
25  (0, 1, 2, 3, 5, 7, 8, 9, 10, 11, 12, 13, 14, 1...   
24  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 15, ...   
23  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 15, ...   
22  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 15, ...   
21  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 15, ...   
20  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 15, ...   
19  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 15, ...   
18  (0, 1, 2, 3, 5, 7, 9, 10, 11, 12, 13, 14, 16, ...   
17  (0, 1, 2, 3, 5, 9, 10, 11, 12, 13, 14, 16, 18,...   
16  (0, 1, 2, 5, 9, 10, 11, 12, 13, 14, 16, 18, 19...   
15  (0, 1, 2, 5, 9, 10, 11, 12, 13, 14, 16, 18, 19...   
14  (0, 1, 2, 5, 9, 10, 11, 12, 13, 14, 16, 18, 19...   
13    (0, 1, 2, 5, 9, 10, 11, 12, 13, 14, 16, 18, 19)   
12        (0, 1, 2, 5, 9, 10, 11, 12, 13, 14, 16, 18)   
11            (0, 1, 2, 5, 9, 10, 11, 12, 13, 14, 16)   
10                (0, 1, 2, 5, 9, 11, 12, 13, 14, 16)   
9                     (0, 1, 2, 5, 9, 11, 12, 13, 14)   
8                         (0, 1, 2, 5, 9, 11, 12, 13)   
7                             (0, 1, 2, 5, 9, 11, 12)   
6                                 (0, 1, 2, 5, 9, 11)   
5                                     (0, 1, 2, 5, 9)   
4                                        (0, 1, 2, 9)   
3                                           (0, 1, 9)   
2                                              (0, 9)   
1                                                (0,)   

                                            cv_scores avg_score  \
27  [0.8863244170752315, 0.827205071966563, 0.8733...  0.868367   
26  [0.8844535636409472, 0.8395809178729241, 0.872...  0.870093   
25  [0.8854663010005633, 0.8407705055712051, 0.872...  0.870637   
24  [0.8868613940392011, 0.8414614409460225, 0.872...   0.87095   
23  [0.8868613940392011, 0.8414614409460224, 0.872...   0.87095   
22  [0.8873092387575998, 0.8414739656024214, 0.872...  0.871045   
21  [0.8870230958096587, 0.842110169737685, 0.8725...  0.871203   
20  [0.8868814591467499, 0.8422885101750054, 0.872...  0.871437   
19  [0.8868897681822827, 0.8422730867222792, 0.872...  0.871489   
18  [0.8868897681822828, 0.8422730867222792, 0.872...  0.871489   
17  [0.8868897681822828, 0.8422730867222792, 0.872...  0.871489   
16  [0.886937132806211, 0.8417821568561511, 0.8718...  0.871302   
15  [0.8871785041347496, 0.8405364753290783, 0.871...  0.871021   
14  [0.8874442225315381, 0.8395216213237058, 0.871...  0.870512   
13  [0.8855293842188675, 0.8376251689853649, 0.872...  0.869996   
12  [0.8849922882828418, 0.8369635400356246, 0.873...  0.869722   
11  [0.8833822096838134, 0.8356489539445545, 0.869...  0.868875   
10  [0.8818206831050115, 0.833604845082666, 0.8685...  0.867832   
9   [0.8773516006399763, 0.8231064203982692, 0.862...  0.865688   
8   [0.8739608483788806, 0.8225777547049642, 0.858...  0.864417   
7   [0.8711717821151814, 0.8200765860514128, 0.852...  0.862976   
6   [0.8692647883836426, 0.816441119930167, 0.8502...  0.860757   
5   [0.8685218034303288, 0.8098226298380534, 0.849...  0.858812   
4   [0.8591903620338608, 0.8306046098225779, 0.840...  0.855734   
3   [0.8552431427246281, 0.821287671646128, 0.7986...  0.837865   
2   [0.7956436929415582, 0.787270788055537, 0.7456...  0.795394   
1   [0.7349323689326035, 0.761324938607378, 0.7043...  0.752469   

                                        feature_names  ci_bound   std_dev  \
27  (Age_08_04, KM, HP, Met_Color, Automatic, CC, ...  0.027283  0.021227   
26  (Age_08_04, KM, HP, Met_Color, Automatic, CC, ...  0.020377  0.015854   
25  (Age_08_04, KM, HP, Met_Color, CC, Cylinders, ...  0.020021  0.015577   
24  (Age_08_04, KM, HP, Met_Color, CC, Cylinders, ...  0.019975  0.015541   
23  (Age_08_04, KM, HP, Met_Color, CC, Cylinders, ...  0.019975  0.015541   
22  (Age_08_04, KM, HP, Met_Color, CC, Cylinders, ...   0.02009  0.015631   
21  (Age_08_04, KM, HP, Met_C

In [305]:
X3b = TDF[['Age_08_04', 'KM', 'Weight']].values
y3b = TDF['Price'].values

In [306]:
X_train, X_test, y_train, y_test = train_test_split(X3b, y3b, test_size=0.25, random_state=3)

In [307]:
linear_model3b=LinearRegression()
linear_model3b.fit(X_train, y_train)

LinearRegression()

In [308]:
# Evaluate Performance
# Training Performance
regressionSummary(y_train, linear_model3b.predict(X_train))

# Test Performance
regressionSummary(y_test, linear_model3b.predict(X_test))


Regression statistics

                      Mean Error (ME) : -0.0000
       Root Mean Squared Error (RMSE) : 1431.4843
            Mean Absolute Error (MAE) : 1025.7794
          Mean Percentage Error (MPE) : -1.3714
Mean Absolute Percentage Error (MAPE) : 10.2303

Regression statistics

                      Mean Error (ME) : -250.3315
       Root Mean Squared Error (RMSE) : 1381.9256
            Mean Absolute Error (MAE) : 987.3214
          Mean Percentage Error (MPE) : -3.1253
Mean Absolute Percentage Error (MAPE) : 9.9057


In [309]:
y_pred_train = linear_model3b.predict(X_train)
R2_train = r2_score(y_train, y_pred_train)

y_pred_test = linear_model3b.predict(X_test)
R2_test = r2_score(y_test, y_pred_test)

print('R2 for training dataset',R2_train)
print('R2 for testing dataset',R2_test)

R2 for training dataset 0.852073868009046
R2 for testing dataset 0.8267851083974975


In [310]:
Complex_Model3a = smf.ols('Price ~ Age_08_04 + KM + Weight', data = TDF).fit()
Complex_Model3a.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Price   R-squared:                       0.848
Model:                            OLS   Adj. R-squared:                  0.848
Method:                 Least Squares   F-statistic:                     2665.
Date:                Thu, 03 Nov 2022   Prob (F-statistic):               0.00
Time:                        13:00:47   Log-Likelihood:                -12454.
No. Observations:                1436   AIC:                         2.492e+04
Df Residuals:                    1432   BIC:                         2.494e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -1880.3356    962.718     -1.953      0.051   -3768.825       8.153
Age_08_04   -120.2212      2.742    -43.841      0.000    -125.600    -114.842
KM            -0.0242      0.001    -20.142      0.000      -0.027      -0.022
Weight        19.5760      0.836     23.409      0.000      17.936      21.216
==============================================================================
Omnibus:                      221.061   Durbin-Watson:                   1.523
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2197.082
Skew:                          -0.373   Prob(JB):                         0.00
Kurtosis:                       9.013   Cond. No.                     2.01e+06
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.01e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [317]:
print('intercept = ', linear_model3b.intercept_) # Printing the interecept for linear regression
print('coefficients', linear_model3b.coef_ )   # Printing the coefficients for linear regression

intercept =  -4349.485791578851
coefficients [-1.19105237e+02 -2.48810345e-02  2.19231205e+01]


In [318]:
print ('Model Form 2=>','Price =', linear_model4.intercept_, '-',1.19105237e+02 ,'* Age_08_04','-', 2.48810345e-02,'* KM', '+', 2.19231205e+01, '* Weight')

Model Form 2=> Price = 36.69290310118993 - 119.105237 * Age_08_04 - 0.0248810345 * KM + 21.9231205 * Weight


# Take note of any differences in model performance from 1. to 2.




MAPE for Model 2 is less than MAPE for model 1, Model 2 performs better than one. 


# Do you notice any major changes in the magnitudes of your parameter estimates?  




Considering Age_08_04 feature there is great difference in the change in the magnitudes, for model 1, it is ~171 & model 2 it is  ~ 119.

# Pick one parameter estimate and, in words, describe what it means?



Considering Age_08_04 here it is negatively correlated to the Price value, so Age_08_04 influences inversly on the Price values, as Age_08_04 increses MEDV value decreases.
Older the car less pricey it is, or the price decreases.

# Question 3 

In [323]:
TDF.columns


Index(['Price', 'Age_08_04', 'KM', 'HP', 'Met_Color', 'Automatic', 'CC',
       'Doors', 'Cylinders', 'Gears', 'Weight', 'Mfr_Guarantee',
       'BOVAG_Guarantee', 'Guarantee_Period', 'Airco', 'Fuel_Type_CNG',
       'Fuel_Type_Diesel', 'Fuel_Type_Petrol', 'Color_Beige', 'Color_Black',
       'Color_Blue', 'Color_Green', 'Color_Grey', 'Color_Red', 'Color_Silver',
       'Color_Violet', 'Color_White', 'Color_Yellow'],
      dtype='object')

In [324]:
TDF.head()

Price  Age_08_04     KM  HP  Met_Color  Automatic    CC  Doors  Cylinders  \
0  13500         23  46986  90          1          0  2000      3          4   
1  13750         23  72937  90          1          0  2000      3          4   
2  13950         24  41711  90          1          0  2000      3          4   
3  14950         26  48000  90          0          0  2000      3          4   
4  13750         30  38500  90          0          0  2000      3          4   

   Gears  ...  Color_Beige  Color_Black  Color_Blue  Color_Green  Color_Grey  \
0      5  ...            0            0           1            0           0   
1      5  ...            0            0           0            0           0   
2      5  ...            0            0           1            0           0   
3      5  ...            0            1           0            0           0   
4      5  ...            0            1           0            0           0   

   Color_Red  Color_Silver  Color_Violet  Color_White  Color_Yellow  
0          0             0             0            0             0  
1          0             1             0            0             0  
2          0             0             0            0             0  
3          0             0             0            0             0  
4          0             0             0            0             0  

[5 rows x 28 columns]

In [334]:
# Selecting necesaary columns only
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, RidgeCV,LassoCV
from sklearn.metrics import mean_absolute_percentage_error, mean_absolute_error, r2_score, mean_squared_error
predictors = ['Age_08_04', 'KM', 'HP', 'Automatic', 'CC', 'Cylinders', 'Gears', 'Weight', 'Mfr_Guarantee',
       'BOVAG_Guarantee', 'Airco', 'Fuel_Type_CNG',
       'Fuel_Type_Diesel', 'Fuel_Type_Petrol'
             ]
X = TDF[predictors]
y = TDF['Price']

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=.25,random_state=1)

r_alphas = [0.04,0.06,0.08,0.085,0.09,0.099,.1,0.11,0.2,0.25]

model = RidgeCV(alphas=r_alphas,normalize=True)
model.fit(X_train,y_train)

train_test_metrics(X_train,X_test,y_train,y_test,model)

           Predictor  coefficent
0          Age_08_04    -112.910
1                 KM      -0.019
2                 HP      47.252
3          Automatic     357.852
4                 CC      -2.689
5          Cylinders       0.000
6              Gears     373.637
7             Weight      18.127
8      Mfr_Guarantee     222.435
9    BOVAG_Guarantee     337.107
10             Airco     451.276
11     Fuel_Type_CNG   -1851.140
12  Fuel_Type_Diesel     867.234
13  Fuel_Type_Petrol    -577.389


             Train          Test
R2    8.670000e-01  5.450000e-01
MAE   9.500540e+02  1.029497e+03
MAPE  9.300000e+00  9.700000e+00
SSE   1.886870e+09  2.138015e+09




C:\Users\anagbhid\Anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), _RidgeGCV())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alphas to: original_alphas * n_samples. 



In [335]:
print('LassoCV Alpha Selection: ',model.alpha_)


LassoCV Alpha Selection:  0.04


In [340]:
print ('Model Form 1 =>','Price =', linear_model3a.intercept_, '+', linear_model3a.coef_, '* Age_08_04')
print()
print ('Model Form 2=>','Price =', linear_model4.intercept_, '-',1.19105237e+02 ,'* Age_08_04','-', 2.48810345e-02,'* KM', 
       '+', 2.19231205e+01, '* Weight')
print()
print ('Model Form 3=>','Price =', model.intercept_, '-',112.910 ,'* Age_08_04','-', 0.019,'* KM', '+',47.252, '* HP', '+', 
       357.852, '* Automatic', '-', 2.689, '* CC', 373.637, '+','* Gears', 18.127, '+','* Weight', 222.435, 
       '*Mfr_Guarantee', 337.107, '+', '* BOVAG_Guarantee', 451.276, '+', '* Airco', '-', 1851.140, '* Fuel_Type_CNG', 
       867.234, '+', '* Fuel_Type_Diesel', '-', 577.389, '* Fuel_Type_Petrol' )

print()
print("Mean Absolute Percentage Error (MAPE) for model 1 : 12.2215")
print("Mean Absolute Percentage Error (MAPE) for model 2 : 9.9057")
print("Mean Absolute Percentage Error (MAPE) for model 3 : 9.7000")

Model Form 1 => Price = 20400.557220502684 + [-171.87956797] * Age_08_04

Model Form 2=> Price = 36.69290310118993 - 119.105237 * Age_08_04 - 0.0248810345 * KM + 21.9231205 * Weight

Model Form 3=> Price = -3806.5227196606593 - 112.91 * Age_08_04 - 0.019 * KM + 47.252 * HP + 357.852 * Automatic - 2.689 * CC 373.637 + * Gears 18.127 + * Weight 222.435 *Mfr_Guarantee 337.107 + * BOVAG_Guarantee 451.276 + * Airco - 1851.14 * Fuel_Type_CNG 867.234 + * Fuel_Type_Diesel - 577.389 * Fuel_Type_Petrol

Mean Absolute Percentage Error (MAPE) for model 1 : 12.2215
Mean Absolute Percentage Error (MAPE) for model 2 : 9.9057
Mean Absolute Percentage Error (MAPE) for model 3 : 9.7000
